jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -128,95 +128,116 @@ public interface HttpChannel extends Invocable
      * <p>Listener instances that are set as a bean on the {@link Connector} are
      * efficiently added to {@link HttpChannel}.
      */
-    public interface Listener extends EventListener
+    interface Listener extends EventListener
     {
         /**
-         * Invoked just after the HTTP request line and headers have been parsed.
+         * Invoked just after the HTTP request line and headers have been parsed
+         * (i.e. from within the call to {@link HttpChannel#onRequest(MetaData.Request)}).
          *
          * <p>
          * This the state of the request from the network, and does not include
          * any request customizations (eg: forwarding, secure, etc)
          * </p>
          *
-         * <p>
-         * Implementations must not block!
-         * </p>
-         *
-         * @param request the request object
+         * @param request the request object. The {@code read()} and {@code demand(Runnable)} methods must not be called by the listener.
```

Comment:
I think calling `read()` and `demand()` should not be a worry -- if called wrongly it's the user fault.